x-pack/plugins/fleet/server/services/managed_package_policies.ts

```diff
@@ -26,78 +30,86 @@ export interface UpgradeManagedPackagePoliciesResult {
-      const dryRunResults = await packagePolicyService.getUpgradeDryRunDiff(
+  for (const { attributes: installedPackage } of installedPackages.saved_objects) {
+    if (installedPackage.keep_policies_up_to_date) {
+      const packagePolicies = await getPackagePoliciesNotMatchingVersion(
         soClient,
-        packagePolicyId
+        installedPackage.name,
+        installedPackage.version
       );
 
-      if (dryRunResults.hasErrors) {
-        const errors = dryRunResults.diff
-          ? dryRunResults.diff?.[1].errors
-          : [dryRunResults.body?.message];
-
-        appContextService
-          .getLogger()
-          .error(
-            new Error(
-              `Error upgrading package policy ${packagePolicyId}: ${JSON.stringify(errors)}`
-            )
-          );
-
-        results.push({ packagePolicyId, diff: dryRunResults.diff, errors });
-        continue;
-      }
-
-      try {
-        await packagePolicyService.upgrade(soClient, esClient, [packagePolicyId]);
-        results.push({ packagePolicyId, diff: dryRunResults.diff, errors: [] });
-      } catch (error) {
-        results.push({ packagePolicyId, diff: dryRunResults.diff, errors: [error] });
+      for (const packagePolicy of packagePolicies) {
+        if (isPolicyVersionLtInstalledVersion(packagePolicy, installedPackage)) {
+          await upgradePackagePolicy(soClient, esClient, packagePolicy.id, results);
+        }
       }
     }
   }
-
   return results;
 };
 
-export function shouldUpgradePolicies(
-  packagePolicyPackageVersion: string,
+async function getPackagePoliciesNotMatchingVersion(
+  soClient: SavedObjectsClientContract,
+  pkgName: string,
+  pkgVersion: string
+): Promise<PackagePolicy[]> {
+  return (
+    await packagePolicyService.list(soClient, {
+      page: 1,
+      perPage: 1000,
+      kuery: `${PACKAGE_POLICY_SAVED_OBJECT_TYPE}.package.name:${pkgName} AND NOT ${PACKAGE_POLICY_SAVED_OBJECT_TYPE}.package.version:${pkgVersion}`,
+    })
+  ).items;
+}
+
+function isPolicyVersionLtInstalledVersion(
+  packagePolicy: PackagePolicy,
   installedPackage: Installation
 ): boolean {
-  const isPolicyVersionGteInstalledVersion = semverGte(
-    packagePolicyPackageVersion,
-    installedPackage.version
+  return (
+    packagePolicy.package !== undefined &&
+    semverLt(packagePolicy.package.version, installedPackage.version)
   );
+}
 
-  return !isPolicyVersionGteInstalledVersion && !!installedPackage.keep_policies_up_to_date;
+async function upgradePackagePolicy(
+  soClient: SavedObjectsClientContract,
+  esClient: ElasticsearchClient,
+  packagePolicyId: string,
+  results: UpgradeManagedPackagePoliciesResult[]
+) {
+  // Since upgrades don't report diffs/errors, we need to perform a dry run first in order
+  // to notify the user of any granular policy upgrade errors that occur during Fleet's
+  // preconfiguration check
+  const dryRunResults = await packagePolicyService.getUpgradeDryRunDiff(soClient, packagePolicyId);
```

Comment:
Worth noting that this upgrade function is not the main pain-point for users since this should only happen once on startup and not really need to be processed again. For users with 1000s of package policies, the improvements in this PR and https://github.com/elastic/kibana/pull/125788 should already be a big gain.

Reply:
Good points, for this pr I will keep the original scope and related tests.

I can raise another pr to take a stab at optimizing `getUpgradeDryRunDiff`. One thing I noticed that it is used from the handler of dry run API too, where those variables are not loaded. So it has to be separated to 2 functions to optimize for setup use case.

Reply:
> So it has to be separated to 2 functions to optimize for setup use case.

Yeah it'd probably make sense to have a function that only does the processing given the already loaded objects and then a function that does the loading and calls the first function.